File: leetcode/0141.linked-list-cycle/go/001.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
// Use fast, slow pointers to determine whether there's a cycle in the linked list.
// Fast pointer moves 2 steps a time, slow pointer moves 1 step a time.
// If fast pointer catch up slow pointer, then there's a cycle in the linked list.
func hasCycle(head *ListNode) bool {
    if head == nil {
        return false
    }

    slow := head.Next
    if slow == nil {
        return false
    }

    fast := slow.Next

    for fast != nil && slow != nil {
        if fast == slow {
            return true
        }

        slow = slow.Next
        
        fast = fast.Next
        if fast != nil {
            fast = fast.Next
        }
    }

    return false
}

File: leetcode/0205.reverse-linked-list/javascript/001.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
const reverseList = function(head) {
  let prev = null;
  let cur = head;

  while(cur) {
    let tmp = cur.next;
    cur.next = prev;
    prev = cur;
    cur = tmp;
  }

  return prev;
};

// Time: O(n)
// Space O(1)